web/source/osk/oskKey.ts

```diff
@@ -169,14 +169,63 @@ namespace com.keyman.osk {
       }
 
       // Apply an overriding class for 5-row layouts
-      var nRows=vkbd.layout['layer'][0]['row'].length;
+      var nRows=vkbd.kbdLayout['layer'][0]['row'].length;
       if(nRows > 4 && vkbd.device.formFactor == 'phone') {
         btn.className='kmw-key kmw-5rows kmw-key-'+OSKKey.BUTTON_CLASSES[n];
       } else {
         btn.className='kmw-key kmw-key-'+OSKKey.BUTTON_CLASSES[n];
       }
     }
 
+    /**
+     * For keys with button classes that support toggle states, this method
+     * may be used to toggle which state the key's button class is in.
+     * -  shift  <=>  shift-on
+     * - special <=> special-on
+     * @param {boolean=} flag The new toggle state 
+     */
+    public setToggleState(vkbd: VisualKeyboard, flag?: boolean) {
+      let btnClassId: number;
+      let classAsString: boolean;
+
+      if(classAsString = typeof this.spec['sp'] == 'string') {
+        btnClassId = parseInt(this.spec['sp'], 10);
+      } else {
+        btnClassId = this.spec['sp'];
+      }
+      
+      // 1 + 2:   shift  +  shift-on
+      // 3 + 4:  special + special-on
+      switch(OSKKey.BUTTON_CLASSES[btnClassId]) {
+        case 'shift':
+        case 'shift-on':
+          if(flag === undefined) {
+            flag = OSKKey.BUTTON_CLASSES[btnClassId] == 'shift';
+          }
+
+          this.spec['sp'] = 1 + (flag ? 1 : 0);
+          break;
+        // New functionality:  we made classes for this but never used 'em!
```

Comment:
Will replace with:

```
        // Added in 15.0:  special key highlight toggling.
        // Was _intended_ in earlier versions, but not actually implemented.
```

That 'special-on' class has existed since KMW 2.0, but I don't see evidence of its actual use in key highlight toggling in prior versions.  That 'highlight toggling' would be in a manner like 'shift' & 'shift-on', which it directly parallels, and which has existed since KMW 2.0.